sql: support `#typeHints` greater than `#placeholders` for prepare stmt

ZhouXing19 opened the pull request:
Previous, we only support pgwire prepare stmt with the number of typehints equal or smaller than the number of the placeholders in the query. E.g. the following usage are not supported:

```
Parse {"Name": "s2", "Query": "select $1", "ParameterOIDs":[1043, 1043, 1043]}
```
Where there is 1 placeholder in the query, but 3 type hints.

This commit is to allow mismatching #typeHints and #placeholders. The former can be larger than the latter now.

This feature is needed for [CRDB's support for Hasura GraphQL Engine](https://github.com/hasura/graphql-engine/issues/8839#issuecomment-1236691642).

Release justification: Low risk, high benefit changes to existing functionality

Release note: For pgwire-level prepare statements, add support for the case where the number of the type hints is greater than the number of placeholders in the given query.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87535)
<!-- Reviewable:end -->

ZhouXing19 commented:
The failed tests in CI don't seem related so RFAL!

rafiss commented on pkg/acceptance/testdata/node/base-test.js:
```diff
@@ -38,11 +38,6 @@ describe('select', () => {
 
 describe('error cases', () => {
   const cases = [{
-    name: 'not enough params',
```
did you confirm that postgres does not expect this behavior?

ZhouXing19 replied:
Added a test [here](https://github.com/cockroachdb/cockroach/pull/87535/commits/4e42e72f4bc7f87b8e6916a7d8b5b293899d0bea#diff-0c2048a8b4d6014b3c805c033e13e0cd30ea5dfb36a3db85d37d34bcc84c4f47R30-R44) to verify.

rafiss reviewed (changes requested):
code looks good! just had testing questions

ZhouXing19 commented on pkg/sql/pgwire/testdata/pgtest/prepare:
```diff
@@ -0,0 +1,32 @@
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"winnie"},{"text":"123"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
```
Done

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87535)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[`pkg/acceptance/testdata/node/base-test.js` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87535#-NBThbhF3c8YB5bfazva:-NBThbhH9ytLOjez_t8f:b-1upxvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/819577e6cd24ca78be5a15f4acdb28264214a02d/pkg/acceptance/testdata/node/base-test.js#L42)):*
> ```JavaScript
>   const cases = [{
>     name: 'not enough params',
>     query: { text: 'SELECT 3', values: ['foo'] },
> ```

hm actually i tried running this against postgres (using the node-pg driver), and it does fail. i think this is a pgtest to reproduce it:

```
Parse {"Name": "s", "Query": "select 3"}
Bind {"DestinationPortal": "p", "PreparedStatement": "s", "ParameterFormatCodes": [0], "Parameters": [{"text":"foo"}]}
Describe {"ObjectType": "P", "Name": "p"}
Execute {"Portal": "p"}
Sync
```
___
*[`pkg/sql/conn_executor_prepare.go` line 389 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87535#-NBT5YPf-hc2ah6yIVLQ:-NBT5YPf-hc2ah6yIVLR:b-i1z9yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e42e72f4bc7f87b8e6916a7d8b5b293899d0bea/pkg/sql/conn_executor_prepare.go#L389)):*
> ```Go
> 			return retErr(
> 				pgwirebase.NewProtocolViolationErrorf(
> 					"expected %d arguments, got %d", numQArgs, len(bindCmd.Args)))
> ```

i see in revision 2 this was added back in. are there already tests for it?


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: support #typeHints greater than #placeholders for prepare stmt

Previous, we only support pgwire prepare stmt with the number of typehints
equal or smaller than the number of the placeholders in the query. E.g. the
following usage are not supported:

```
Parse {"Name": "s2", "Query": "select $1", "ParameterOIDs":[1043, 1043, 1043]}
```
Where there are 1 placeholder in the query, but 3 type hints.

This commit is to allow mismatching #typeHints and #placeholders. The former
can be larger than the latter now.

Release justification: Low risk, high benefit changes to existing functionality

Release note (sql change): For pgwire-level prepare statements, support the case
where the number of the type hints is greater than the number of placeholders
in the given query.

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87535)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/conn_executor_prepare.go` line 389 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87535#-NBT5YPf-hc2ah6yIVLQ:-NC7OVOP76gfMoafpytm:b-u444km) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e42e72f4bc7f87b8e6916a7d8b5b293899d0bea/pkg/sql/conn_executor_prepare.go#L389)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i see in revision 2 this was added back in. are there already tests for it?
</blockquote></details>

Yes, it was because removing this part made this test pass:
```
Bind {"DestinationPortal": "p2", "PreparedStatement": "s2", "ParameterFormatCodes": [0], "Parameters": [{"text":"whitebear"}]}
Execute {"Portal": "p2"}
Sync
```
which is incorrect. I've added this test to the pgtest too.
___
*[`pkg/acceptance/testdata/node/base-test.js` line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87535#-NBThbhF3c8YB5bfazva:-NC7OFSz0LnOeWMqqJ56:bdrbn55) ([raw file](https://github.com/cockroachdb/cockroach/blob/819577e6cd24ca78be5a15f4acdb28264214a02d/pkg/acceptance/testdata/node/base-test.js#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm actually i tried running this against postgres (using the node-pg driver), and it does fail. i think this is a pgtest to reproduce it:

```
Parse {"Name": "s", "Query": "select 3"}
Bind {"DestinationPortal": "p", "PreparedStatement": "s", "ParameterFormatCodes": [0], "Parameters": [{"text":"foo"}]}
Describe {"ObjectType": "P", "Name": "p"}
Execute {"Portal": "p"}
Sync
```
</blockquote></details>

Yeah good point, I added this one to the pgtest, and put back what we had here.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for the fix!

> Reviewed 11 of 14 files at r1, 1 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87535#-:-NCFvXxh-Dus2J7sU5Hs:bp9cthg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r=rafiss
